<!doctype html><html lang=en-us><head><meta charset=utf-8><meta http-equiv=x-ua-compatible content="ie=edge"><meta name=viewport content="width=device-width,initial-scale=1,shrink-to-fit=no"><link rel=preload as=font href=fonts/vendor/jost/jost-v4-latin-regular.woff2 type=font/woff2 crossorigin><link rel=preload as=font href=fonts/vendor/jost/jost-v4-latin-500.woff2 type=font/woff2 crossorigin><link rel=preload as=font href=fonts/vendor/jost/jost-v4-latin-700.woff2 type=font/woff2 crossorigin><link rel=stylesheet href=/main.568d041fc4cd80ef2f40f614591cd4edce13e9a07325a64d67ebd0cc9a6074daef6c901632a7b7163cc23fa99339b7d5ef55b8329dfa6fc5003554242a140bdd.css integrity="sha512-Vo0EH8TNgO8vQPYUWRzU7c4T6aBzJaZNZ+vQzJpgdNrvbJAWMqe3FjzCP6mTObfV71W4Mp36b8UANVQkKhQL3Q==" crossorigin=anonymous><noscript><style>img.lazyload{display:none}</style></noscript><meta name=robots content="index, follow"><meta name=googlebot content="index, follow, max-snippet:-1, max-image-preview:large, max-video-preview:-1"><meta name=bingbot content="index, follow, max-snippet:-1, max-image-preview:large, max-video-preview:-1"><title>Using Texera to Perform Single-cell RNA Sequencing Analysis with R Language - Texera</title><meta name=description content="RNA sequence analysis has revolutionized our understanding of gene expression at the single-cell level, providing unprecedented insights into cellular heterogeneity and dynamics. Seurat, a popular R library, has emerged as a versatile toolkit for the comprehensive analysis of single-cell RNA sequencing (scRNA-seq) data.
While the core engine of Texera is built using Scala, it supports the integration of Python UDFs, allowing users to incorporate custom Python code directly into their data processing pipelines."><link rel=canonical href=/blog/using-texera-to-perform-single-cell-rna-sequencing-analysis-with-r-language/><meta property="og:locale" content="en_US"><meta property="og:type" content="article"><meta property="og:title" content="Using Texera to Perform Single-cell RNA Sequencing Analysis with R Language"><meta property="og:description" content="RNA sequence analysis has revolutionized our understanding of gene expression at the single-cell level, providing unprecedented insights into cellular heterogeneity and dynamics. Seurat, a popular R library, has emerged as a versatile toolkit for the comprehensive analysis of single-cell RNA sequencing (scRNA-seq) data.
While the core engine of Texera is built using Scala, it supports the integration of Python UDFs, allowing users to incorporate custom Python code directly into their data processing pipelines."><meta property="og:url" content="/blog/using-texera-to-perform-single-cell-rna-sequencing-analysis-with-r-language/"><meta property="og:site_name" content="Texera"><meta property="article:published_time" content="2023-06-27T14:26:21-07:00"><meta property="article:modified_time" content="2023-06-27T14:26:21-07:00"><meta property="og:image" content="doks.png"><meta property="og:image:alt" content="Texera"><meta name=twitter:card content="summary_large_image"><meta name=twitter:site content><meta name=twitter:creator content><meta name=twitter:title content="Using Texera to Perform Single-cell RNA Sequencing Analysis with R Language"><meta name=twitter:description content><meta name=twitter:image content="doks.png"><meta name=twitter:image:alt content="Using Texera to Perform Single-cell RNA Sequencing Analysis with R Language"><script type=application/ld+json>{"@context":"https://schema.org","@graph":[{"@type":"Organization","@id":"/#/schema/organization/1","name":"Texera","url":"/","sameAs":["https://github.com/texera/texera"],"logo":{"@type":"ImageObject","@id":"/#/schema/image/1","url":"/logo-doks.png","width":512,"height":512,"caption":"Texera"},"image":{"@id":"/#/schema/image/1"}},{"@type":"WebSite","@id":"/#/schema/website/1","url":"/","name":"Texera","description":"Texera is a system to support collaborative, ML-centric data analytics as a cloud-based service using GUI-based workflows. It supports scalable computation with a parallel backend engine, and enables advanced AI\/ML techniques.","publisher":{"@id":"/#/schema/organization/1"}},{"@type":"WebPage","@id":"/blog/using-texera-to-perform-single-cell-rna-sequencing-analysis-with-r-language/","url":"/blog/using-texera-to-perform-single-cell-rna-sequencing-analysis-with-r-language/","name":"Using Texera to Perform Single-cell RNA Sequencing Analysis with R Language","description":"","isPartOf":{"@id":"/#/schema/website/1"},"about":{"@id":"/#/schema/organization/1"},"datePublished":"2023-06-27T14:26:21CET","dateModified":"2023-06-27T14:26:21CET","breadcrumb":{"@id":"/blog/using-texera-to-perform-single-cell-rna-sequencing-analysis-with-r-language/#/schema/breadcrumb/1"},"primaryImageOfPage":{"@id":"/blog/using-texera-to-perform-single-cell-rna-sequencing-analysis-with-r-language/#/schema/image/2"},"inLanguage":"en-US","potentialAction":[{"@type":"ReadAction","target":["/blog/using-texera-to-perform-single-cell-rna-sequencing-analysis-with-r-language/"]}]},{"@type":"BreadcrumbList","@id":"/blog/using-texera-to-perform-single-cell-rna-sequencing-analysis-with-r-language/#/schema/breadcrumb/1","name":"Breadcrumbs","itemListElement":[{"@type":"ListItem","position":1,"item":{"@type":"WebPage","@id":"","url":"","name":"Home"}},{"@type":"ListItem","position":3,"item":{"@type":"WebPage","@id":"/blog/","url":"/blog/","name":"Blog"}},{"@type":"ListItem","position":4,"item":{"@id":"/blog/using-texera-to-perform-single-cell-rna-sequencing-analysis-with-r-language/"}}]},{"@context":"https://schema.org","@graph":[{"@type":"Article","@id":"/#/schema/article/1","headline":"Using Texera to Perform Single-cell RNA Sequencing Analysis with R Language","description":"","isPartOf":{"@id":"/blog/using-texera-to-perform-single-cell-rna-sequencing-analysis-with-r-language/"},"mainEntityOfPage":{"@id":"/blog/using-texera-to-perform-single-cell-rna-sequencing-analysis-with-r-language/"},"datePublished":"2023-06-27T14:26:21CET","dateModified":"2023-06-27T14:26:21CET","author":{"@id":"/#/schema/person/2"},"publisher":{"@id":"/#/schema/organization/1"},"image":{"@id":"/blog/using-texera-to-perform-single-cell-rna-sequencing-analysis-with-r-language/#/schema/image/2"}}]},{"@context":"https://schema.org","@graph":[{"@type":"Person","@id":"/#/schema/person/2","name":"Texera Team","sameAs":[]}]},{"@context":"https://schema.org","@graph":[{"@type":"ImageObject","@id":"/blog/using-texera-to-perform-single-cell-rna-sequencing-analysis-with-r-language/#/schema/image/2","url":"doks.png","contentUrl":"doks.png","caption":"Using Texera to Perform Single-cell RNA Sequencing Analysis with R Language"}]}]}</script><meta name=theme-color content="#fff"><link rel=apple-touch-icon sizes=180x180 href=apple-touch-icon.png><link rel=icon type=image/png sizes=32x32 href=favicon-32x32.png><link rel=icon type=image/png sizes=16x16 href=favicon-16x16.png><link rel=manifest crossorigin=use-credentials href=site.webmanifest></head><body class="blog single"><div class=header-bar></div><header class="navbar navbar-expand-md navbar-light doks-navbar"><nav class="container-fluid flex-wrap flex-md-nowrap" aria-label="Main navigation"><a class="navbar-brand p-0 me-auto" href=/ aria-label=Texera>Texera</a>
<button class="btn btn-menu d-block d-md-none order-5" type=button data-bs-toggle=offcanvas data-bs-target=#offcanvasDoks aria-controls=offcanvasDoks aria-label="Open main menu"><svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24" fill="none" stroke="currentcolor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-menu"><line x1="3" y1="12" x2="21" y2="12"/><line x1="3" y1="6" x2="21" y2="6"/><line x1="3" y1="18" x2="21" y2="18"/></svg></button><div class="offcanvas offcanvas-end border-0 py-md-1" tabindex=-1 id=offcanvasDoks data-bs-backdrop=true aria-labelledby=offcanvasDoksLabel><div class="header-bar d-md-none"></div><div class="offcanvas-header d-md-none"><h2 class="h5 offcanvas-title ps-2" id=offcanvasDoksLabel><a class=text-dark href=/>Texera</a></h2><button type=button class="btn-close text-reset me-2" data-bs-dismiss=offcanvas aria-label="Close main menu"></button></div><div class="offcanvas-body px-4"><h3 class="h6 text-uppercase mb-3 d-md-none">Main</h3><ul class="nav flex-column flex-md-row ms-md-n3"><li class=nav-item><a class="nav-link ps-0 py-1 active" href=/blog/>Blog</a></li></ul><hr class="text-black-50 my-4 d-md-none"><h3 class="h6 text-uppercase mb-3 d-md-none">Socials</h3><ul class="nav flex-column flex-md-row ms-md-auto me-md-n5 pe-md-2"><li class=nav-item><a class="nav-link ps-0 py-1" href=/https:/github.com/Texera/texera><svg xmlns="http://www.w3.org/2000/svg" width="20" height="20" viewBox="0 0 24 24" fill="none" stroke="currentcolor" stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather feather-github"><path d="M9 19c-5 1.5-5-2.5-7-3m14 6v-3.87a3.37 3.37.0 00-.94-2.61c3.14-.35 6.44-1.54 6.44-7A5.44 5.44.0 0020 4.77 5.07 5.07.0 0019.91 1S18.73.65 16 2.48a13.38 13.38.0 00-7 0C6.27.65 5.09 1 5.09 1A5.07 5.07.0 005 4.77 5.44 5.44.0 003.5 8.55c0 5.42 3.3 6.61 6.44 7A3.37 3.37.0 009 18.13V22"/></svg><small class="ms-2 d-md-none">GitHub</small></a></li></ul></div></div></nav></header><div class="wrap container-fluid" role=document><div class=content><div class="row justify-content-center"><div class="col-md-12 col-lg-10 col-xl-8"><article><div class=blog-header><h1>Using Texera to Perform Single-cell RNA Sequencing Analysis with R Language</h1><p><small>Posted June 27, 2023
by
<a class="stretched-link position-relative" href=/contributors/yicong-huang/>Yicong Huang</a>
from
<a class="stretched-link position-relative" href=/affiliations/computer-science-uc-irvine/>Computer Science, UC Irvine</a>
&nbsp;&dash;&nbsp;
<strong>7&nbsp;min read</strong></small><p></div><p class=lead>We share the experience of using Python UDFs in Texera to execute R code to perform RNA sequence analysis.</p><p>RNA sequence analysis has revolutionized our understanding of gene expression at the single-cell level, providing unprecedented insights into cellular heterogeneity and dynamics. <a href=https://satijalab.org/seurat/index.html>Seurat</a>, a popular R library, has emerged as a versatile toolkit for the comprehensive analysis of single-cell RNA sequencing (<a href=https://www.ncbi.nlm.nih.gov/pmc/articles/PMC8964935/>scRNA-seq</a>) data.</p><p>While the core engine of Texera is built using Scala, it supports the integration of Python UDFs, allowing users to incorporate custom Python code directly into their data processing pipelines.
This capability enables users to leverage the versatility and extensive ecosystem of Python libraries for their data-processing tasks.
However, the Seurat library is primarily available in the R programming language, which poses a challenge when it comes to processing RNA sequence data with Texera, which previously did not support R.</p><p>In this post, we share our experience of overcoming the challenges of processing RNA sequence data with R in Texera by leveraging the platform&rsquo;s support for Python UDFs as a bridge.</p><h3 id=task-overview>Task Overview <a href=#task-overview class=anchor aria-hidden=true>#</a></h3><p>We have an existing script written in the R programming language that performs quality control (QC) on scRNA-seq data from human islets (<a href=https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6609987/>HPAP</a>) across different ages. The dataset includes samples from individuals aged 32, 36, and 53.
The R script loads the dataset, creates a Seurat instance to process the dataset, then generates plots to visualize the results.
The original R script contains 524 lines of code. We want to migrate the R script to run on Texera.</p><details><summary>The R script code (Partial)</summary><div class=highlight><pre tabindex=0 class=chroma><code class=language-r data-lang=r><span class=line><span class=cl><span class=nf>library</span><span class=p>(</span><span class=n>Seurat</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=nf>library</span><span class=p>(</span><span class=n>scater</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=nf>library</span><span class=p>(</span><span class=n>scran</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=nf>library</span><span class=p>(</span><span class=n>batchelor</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=nf>library</span><span class=p>(</span><span class=n>tidyverse</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=c1># sample info</span>
</span></span><span class=line><span class=cl><span class=n>sample.info</span> <span class=o>&lt;-</span> <span class=nf>data.frame</span><span class=p>(</span><span class=n>SeqName</span><span class=o>=</span><span class=nf>c</span><span class=p>(</span><span class=s>&#39;HPAP032&#39;</span><span class=p>,</span><span class=s>&#39;HPAP036&#39;</span><span class=p>,</span><span class=s>&#39;HPAP053&#39;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                         <span class=n>Group</span><span class=o>=</span><span class=nf>c</span><span class=p>(</span><span class=s>&#39;T1D&#39;</span><span class=p>,</span><span class=s>&#39;No Diabetes&#39;</span><span class=p>,</span><span class=s>&#39;No Diabetes&#39;</span><span class=p>),</span>
</span></span><span class=line><span class=cl>                         <span class=n>Name</span><span class=o>=</span><span class=nf>c</span><span class=p>(</span><span class=s>&#39;P32&#39;</span><span class=p>,</span><span class=s>&#39;P36&#39;</span><span class=p>,</span><span class=s>&#39;P53&#39;</span><span class=p>))</span>
</span></span><span class=line><span class=cl><span class=nf>rownames</span><span class=p>(</span><span class=n>sample.info</span><span class=p>)</span> <span class=o>&lt;-</span> <span class=n>sample.info</span><span class=o>$</span><span class=n>Name</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=c1># load corrected UMI counts table per patient</span>
</span></span><span class=line><span class=cl><span class=n>raw.counts.list</span> <span class=o>&lt;-</span> <span class=nf>list</span><span class=p>()</span>
</span></span><span class=line><span class=cl><span class=nf>for </span><span class=p>(</span><span class=n>k</span> <span class=n>in</span> <span class=m>1</span><span class=o>:</span><span class=nf>nrow</span><span class=p>(</span><span class=n>sample.info</span><span class=p>)){</span>
</span></span><span class=line><span class=cl>   <span class=n>pid</span> <span class=o>&lt;-</span> <span class=nf>rownames</span><span class=p>(</span><span class=n>sample.info</span><span class=p>)</span><span class=n>[k]</span>
</span></span><span class=line><span class=cl>   <span class=n>sid</span> <span class=o>&lt;-</span> <span class=n>sample.info</span><span class=o>$</span><span class=n>SeqName[k]</span>
</span></span><span class=line><span class=cl>   <span class=n>raw.counts.list[[k]]</span> <span class=o>&lt;-</span> <span class=nf>my.Read10X</span><span class=p>(</span><span class=nf>file.path</span><span class=p>(</span><span class=n>sourcedir</span><span class=p>,</span> <span class=n>sid</span><span class=p>,</span> <span class=s>&#39;filtered_feature_bc_matrix&#39;</span><span class=p>),</span> <span class=n>pid</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=p>}</span>
</span></span><span class=line><span class=cl><span class=nf>names</span><span class=p>(</span><span class=n>raw.counts.list</span><span class=p>)</span> <span class=o>&lt;-</span> <span class=nf>rownames</span><span class=p>(</span><span class=n>sample.info</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=kc>...</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=c1># Initialize the Seurat object with the raw (non-normalized data).</span>
</span></span><span class=line><span class=cl><span class=n>panc.initial</span> <span class=o>&lt;-</span> <span class=nf>CreateSeuratObject</span><span class=p>(</span><span class=n>counts</span><span class=o>=</span><span class=n>raw.counts.all</span><span class=p>,</span> <span class=n>project</span><span class=o>=</span><span class=n>project</span><span class=p>,</span> <span class=n>assay</span><span class=o>=</span><span class=s>&#34;RNA&#34;</span><span class=p>,</span> <span class=n>min.cells</span><span class=o>=</span><span class=m>0</span><span class=p>,</span> <span class=n>min.features</span><span class=o>=</span><span class=m>0</span><span class=p>,</span>
</span></span><span class=line><span class=cl>                                  <span class=n>names.field</span><span class=o>=</span><span class=m>1</span><span class=p>,</span> <span class=n>names.delim</span><span class=o>=</span><span class=s>&#34;_&#34;</span><span class=p>,</span> <span class=n>meta.data</span><span class=o>=</span><span class=kc>NULL</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=c1># Calculates the mitochondrial/ribosomal genes per cell</span>
</span></span><span class=line><span class=cl><span class=nf>print</span><span class=p>(</span><span class=nf>paste</span><span class=p>(</span><span class=nf>length</span><span class=p>(</span><span class=nf>grep</span><span class=p>(</span><span class=n>mito.pattern</span><span class=p>,</span> <span class=nf>rownames</span><span class=p>(</span><span class=n>panc.initial</span><span class=p>),</span> <span class=n>value</span><span class=o>=</span><span class=bp>T</span><span class=p>)),</span> <span class=s>&#39;mitochondrial genes to consider.&#39;</span><span class=p>))</span>
</span></span><span class=line><span class=cl><span class=nf>print</span><span class=p>(</span><span class=nf>paste</span><span class=p>(</span><span class=nf>length</span><span class=p>(</span><span class=nf>grep</span><span class=p>(</span><span class=n>ribo.pattern</span><span class=p>,</span> <span class=nf>rownames</span><span class=p>(</span><span class=n>panc.initial</span><span class=p>),</span> <span class=n>value</span><span class=o>=</span><span class=bp>T</span><span class=p>)),</span> <span class=s>&#39;ribosomal genes to consider.&#39;</span><span class=p>))</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=n>panc.initial[[</span><span class=s>&#34;percent.mito&#34;</span><span class=n>]]</span> <span class=o>&lt;-</span> <span class=nf>PercentageFeatureSet</span><span class=p>(</span><span class=n>panc.initial</span><span class=p>,</span> <span class=n>pattern</span><span class=o>=</span><span class=n>mito.pattern</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>panc.initial[[</span><span class=s>&#34;percent.ribo&#34;</span><span class=n>]]</span> <span class=o>&lt;-</span> <span class=nf>PercentageFeatureSet</span><span class=p>(</span><span class=n>panc.initial</span><span class=p>,</span> <span class=n>pattern</span><span class=o>=</span><span class=n>ribo.pattern</span><span class=p>)</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=kc>...</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl><span class=c1># scatter plot</span>
</span></span><span class=line><span class=cl><span class=n>plots</span> <span class=o>&lt;-</span> <span class=nf>list</span><span class=p>()</span>
</span></span><span class=line><span class=cl><span class=n>plots[[1]]</span> <span class=o>&lt;-</span> <span class=nf>FeatureScatter</span><span class=p>(</span><span class=n>panc.initial</span><span class=p>,</span> <span class=n>feature1</span><span class=o>=</span><span class=s>&#34;nFeature_RNA&#34;</span><span class=p>,</span> <span class=n>feature2</span><span class=o>=</span><span class=s>&#34;nCount_RNA&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>plots[[2]]</span> <span class=o>&lt;-</span> <span class=nf>FeatureScatter</span><span class=p>(</span><span class=n>panc.initial</span><span class=p>,</span> <span class=n>feature1</span><span class=o>=</span><span class=s>&#34;nCount_RNA&#34;</span><span class=p>,</span> <span class=n>feature2</span><span class=o>=</span><span class=s>&#34;percent.mito&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>plots[[3]]</span> <span class=o>&lt;-</span> <span class=nf>FeatureScatter</span><span class=p>(</span><span class=n>panc.initial</span><span class=p>,</span> <span class=n>feature1</span><span class=o>=</span><span class=s>&#34;nFeature_RNA&#34;</span><span class=p>,</span> <span class=n>feature2</span><span class=o>=</span><span class=s>&#34;percent.ribo&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>plots[[4]]</span> <span class=o>&lt;-</span> <span class=nf>FeatureScatter</span><span class=p>(</span><span class=n>panc.initial</span><span class=p>,</span> <span class=n>feature1</span><span class=o>=</span><span class=s>&#34;percent.mito&#34;</span><span class=p>,</span> <span class=n>feature2</span><span class=o>=</span><span class=s>&#34;percent.ribo&#34;</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=n>plots.combined</span> <span class=o>&lt;-</span> <span class=p>(</span><span class=n>plots[[1]]</span> <span class=o>|</span> <span class=n>plots[[2]]</span><span class=p>)</span> <span class=o>/</span> <span class=p>(</span><span class=n>plots[[3]]</span> <span class=o>|</span> <span class=n>plots[[4]]</span><span class=p>)</span>
</span></span><span class=line><span class=cl><span class=nf>ggsave</span><span class=p>(</span><span class=nf>file.path</span><span class=p>(</span><span class=n>figdir</span><span class=p>,</span> <span class=s>&#34;QC.Scatter.pre-filter.png&#34;</span><span class=p>),</span> <span class=n>width</span><span class=o>=</span><span class=m>9</span><span class=p>,</span> <span class=n>height</span><span class=o>=</span><span class=m>7.5</span><span class=p>,</span> <span class=n>dpi</span><span class=o>=</span><span class=m>300</span><span class=p>)</span>
</span></span></code></pre></div></details><h3 id=initial-migration-of-r-code-in-a-single-python-udf>Initial migration of R code in a single Python UDF <a href=#initial-migration-of-r-code-in-a-single-python-udf class=anchor aria-hidden=true>#</a></h3><p>As a starting point, we took the R script as it is and migrated into Texera with one single UDF. The following Figure 1 shows the initial Texera workflow.</p><figure><a href=single-udf-workflow.png><img src=single-udf-workflow.png alt="Fig 1" style=width:100%></a><figcaption align=center><i>Figure 1. Initial workflow that executes R code in one single Python UDF.</i></figcaption></figure><ul><li><p>The first operator, <code>Seuret Runner (R)</code>, is a Python UDF operator that executes the R source code, and renders the
generated plots as a html.</p></li><li><p>The second operator, <code>HTML Visualizer</code>, visualizes the generated html.</p></li></ul><h4 id=python-udf-code-for-seurat-runner-r>Python UDF code for Seurat Runner (R) <a href=#python-udf-code-for-seurat-runner-r class=anchor aria-hidden=true>#</a></h4><figure><a href=single-udf-code.png><img src=single-udf-code.png alt="Fig 2" style=width:100%></a><figcaption align=center><i>Figure 2. The code of the single UDF which executes the entire R code as it is.</i></figcaption></figure><p>We used a single Python UDF operator to execute R script code. In line 8, the <code>r_code</code> variable takes the R script
source code, which we folded for simplicity, as a string. In line 540, we execute the <code>r_code</code> as its entirety with a
library called &ldquo;<a href=https://rpy2.github.io>rpy2</a>&rdquo;, which will invoke the Seurat library to perform the analysis and generate plots on disk.</p><p>After that, we read the generated plots from disk and put them into an html for visualization.</p><h4 id=executing-r-code-in-python>Executing R code in Python <a href=#executing-r-code-in-python class=anchor aria-hidden=true>#</a></h4><p>In order to execute the R code within Python, we made use of the <code>rpy2</code> Python library.
This library facilitates the execution of R code within Python environments, enabling users to seamlessly call R functions, execute R scripts, and manipulate R objects directly from their Python code.</p><p>Specifically, in this single UDF implementation, we took R code as it is and executed it with:</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-python data-lang=python><span class=line><span class=cl><span class=n>rpy2</span><span class=o>.</span><span class=n>robjects</span><span class=o>.</span><span class=n>r</span><span class=p>[</span><span class=n>r_code</span><span class=p>]</span>
</span></span></code></pre></div><p>The evaluation of the R code took place within the Global Environment, which is the default starting point for R users when they access the R console.
In this environment, any variables created by the R code were automatically stored and becoming accessible for subsequent operations in R.
In this case, the Python UDF simply worked as an invoker of the R code, without any data transformation between the R environment and the Python environment.</p><h4 id=visualizing-plots-as-an-html>Visualizing plots as an HTML <a href=#visualizing-plots-as-an-html class=anchor aria-hidden=true>#</a></h4><p>All the plots are rendered in one HTML file, which can be visualized in Texera. Users can click the &ldquo;view result&rdquo; operator to see the visualization pop up window. Figure 3 shows the Texera UI of rendering the visualized plots in HTML.</p><figure><a href=html-visualization.png><img src=html-visualization.png alt="Fig 3" style=width:100%></a><figcaption align=center><i>Figure 3. Visualized plots in Texera UI as an HTML.</i></figcaption></figure><p>Leveraging the extensive capabilities of HTML, Texera is able to effortlessly display the plots generated. Figure 4 presents several examples of the plots generated by the R code.</p><figure><div style=width:50%;float:left;padding:20px><a href=umap-plot.png><img src=umap-plot.png alt="Fig 4a" style=width:100%></a><figcaption align=center style=width:100%><i>(A) UMAP plot by cell types.</i></figcaption></div><div style=width:50%;float:left;padding:20px><a href=scatterplot.png><img src=scatterplot.png alt="Fig 4b" style=width:100%></a><figcaption align=center style=width:100%><i>(B) Scatter plot by ages.</i></figcaption></div><div style=padding:20px><a href=heatmap-wilcox-plot.png><img src=heatmap-wilcox-plot.png alt="Fig 4c" style=width:100%></a><figcaption align=center style=width:100%><i>(C) Heatmap wilcox plot.</i></figcaption></div><figcaption align=center style=width:100%><i>Figure 4. Some example plots generated by the R code for displaying scRNA sequences.</i></figcaption></figure><h3 id=decomposing-the-single-udf-into-multiple-udfs>Decomposing the single UDF into multiple UDFs <a href=#decomposing-the-single-udf-into-multiple-udfs class=anchor aria-hidden=true>#</a></h3><p>Next, we expanded the single UDF implementation into a more intricate workflow. Figure 5 illustrates the resulting workflow, which provides a more comprehensive and meaningful depiction of the data-processing steps.</p><figure><a href=decomposed-workflow.png><img src=decomposed-workflow.png alt="Fig 5" style=width:100%></a><figcaption align=center><i>Figure 5. The complex workflow after decomposing the single UDF implementation.</i></figcaption></figure><ul><li>The <code>Raw Counts Scan</code> operator scans 3 scRNA datasets from disk and merges them into a two dimensional matrix.</li><li>The <code>Seurat</code> operator creates the Seurat instance with the matrix by invoking the R library.</li><li>The <code>QC</code> operators each generate one corresponding plot and embed it into an HTML string.</li><li>The <code>HTML visualizer</code> operators visualize the generated plots respectively.</li><li>The <code>filter</code> operator applies a filter logic on the Seurat instance.</li><li>The <code>collect stats of filter cells</code> aggregates metadata of the filtered dataset.</li></ul><h4 id=transferring-r-objects-back-to-python>Transferring R objects back to Python <a href=#transferring-r-objects-back-to-python class=anchor aria-hidden=true>#</a></h4><p>In order to decompose the single UDF and utilize the edges to pass data between operators, we face the challenge of passing data from the R environment to the Python environment. <code>rpy2</code> library supports converters from R objects to Python objects.</p><p>For primitive data types such as integer and string, <code>rpy2</code> provides default converters. The R code executed with those converters can be available later in the hosting Python environment. In the following example, the R code assigns a variable <code>panc.initial</code>, with the converter enabled on the Python side:</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-python data-lang=python><span class=line><span class=cl><span class=k>with</span> <span class=n>rpy2</span><span class=o>.</span><span class=n>robjects</span><span class=o>.</span><span class=n>conversion</span><span class=o>.</span><span class=n>localconverter</span><span class=p>(</span><span class=n>rpy2</span><span class=o>.</span><span class=n>robjects</span><span class=o>.</span><span class=n>default_converter</span><span class=p>)</span> <span class=k>as</span> <span class=n>cv</span><span class=p>:</span>
</span></span><span class=line><span class=cl>    <span class=n>rpy2</span><span class=o>.</span><span class=n>robjects</span><span class=o>.</span><span class=n>r</span><span class=p>(</span><span class=s1>&#39;panc.initial &lt;- CreateSeuratObject(counts=raw.counts.all, project=project, assay=&#34;RNA&#34;, min.cells=0, min.features=0, names.field=1, names.delim=&#34;_&#34;, meta.data=NULL)&#39;</span><span class=p>)</span>
</span></span></code></pre></div><p>After the R code executes, we can get the <code>panc.initial</code> instance as a Python object:</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-python data-lang=python><span class=line><span class=cl><span class=n>panc_initial</span> <span class=o>=</span> <span class=n>rpy2</span><span class=o>.</span><span class=n>robjects</span><span class=o>.</span><span class=n>r</span><span class=p>[</span><span class=s1>&#39;panc.initial&#39;</span><span class=p>]</span>
</span></span></code></pre></div><p>For complex custom data structures and types, we can also provide our own converter implementation to accommodate rare use cases.</p><p>With converters, Python UDF in Texera can get the necessary data objects from the R environment back to Python.</p><h4 id=transferring-r-objects-between-operators>Transferring R objects between operators <a href=#transferring-r-objects-between-operators class=anchor aria-hidden=true>#</a></h4><p>Another challenge to support the decomposed workflow is how to transfer the R data between two operators. In Texera, each operator is self-contained and has its individual processing environment. In this case, each Python UDF will have its own Python environment and R environment.</p><p>Consider the <code>Seurat</code> operator in Figure 5. It generates the Seurat instance and needs to share the same instance to five downstream operators, each performing different processing. To do so, we used Texera’s binary type to transfer the <code>Seurat</code> instance.</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-python data-lang=python><span class=line><span class=cl><span class=k>class</span> <span class=nc>SeuratCreator</span><span class=p>(</span><span class=n>UDFTupleOperatorV2</span><span class=p>):</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>    <span class=k>def</span> <span class=nf>process_tuple</span><span class=p>(</span><span class=bp>self</span><span class=p>,</span> <span class=n>tuple_</span><span class=p>:</span> <span class=n>Tuple</span><span class=p>,</span> <span class=n>port</span><span class=p>:</span> <span class=nb>int</span><span class=p>)</span><span class=o>-&gt;</span> <span class=n>Optional</span><span class=p>[</span><span class=n>TupleLike</span><span class=p>]:</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>        <span class=o>...</span> <span class=c1># r code execution with converters enabled</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>        <span class=k>yield</span> <span class=n>Tuple</span><span class=p>({</span><span class=s2>&#34;panc.initial&#34;</span><span class=p>:</span> <span class=n>rpy2</span><span class=o>.</span><span class=n>robjects</span><span class=o>.</span><span class=n>r</span><span class=p>[</span><span class=s1>&#39;panc.initial&#39;</span><span class=p>]})</span> <span class=c1># output as binary type</span>
</span></span></code></pre></div><p>The binary type will be serialized by the Python’s serializer library, pickle, and sent to the downstream operators as bytes. The receiver operators will first deserialize with pickle, and make the instance available in the Python environment. Then, we need to make the instance available in the receiver operator’s R environment, thus we can assign it back to the corresponding R global environment:</p><div class=highlight><pre tabindex=0 class=chroma><code class=language-python data-lang=python><span class=line><span class=cl><span class=k>class</span> <span class=nc>ScatterplotGenerator</span><span class=p>(</span><span class=n>UDFTupleOperatorV2</span><span class=p>):</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>    <span class=k>def</span> <span class=nf>process_tuple</span><span class=p>(</span><span class=bp>self</span><span class=p>,</span> <span class=n>tuple_</span><span class=p>:</span> <span class=n>Tuple</span><span class=p>,</span> <span class=n>port</span><span class=p>:</span> <span class=nb>int</span><span class=p>)</span><span class=o>-&gt;</span> <span class=n>Optional</span><span class=p>[</span><span class=n>TupleLike</span><span class=p>]:</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>        <span class=c1># assign the Seurat instance into the current R environment</span>
</span></span><span class=line><span class=cl>        <span class=n>rpy2</span><span class=o>.</span><span class=n>robjects</span><span class=o>.</span><span class=n>globalenv</span><span class=p>[</span><span class=s1>&#39;panc.initial&#39;</span><span class=p>]</span> <span class=o>=</span> <span class=n>tuple_</span><span class=p>[</span><span class=s1>&#39;panc.initial&#39;</span><span class=p>]</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>        <span class=o>...</span> <span class=c1># use r code to generate plots from the Seurat instance</span>
</span></span><span class=line><span class=cl>
</span></span><span class=line><span class=cl>        <span class=k>yield</span> <span class=n>Tuple</span><span class=p>({</span><span class=s1>&#39;html&#39;</span><span class=p>:</span> <span class=n>rpy2</span><span class=o>.</span><span class=n>robjects</span><span class=o>.</span><span class=n>r</span><span class=p>[</span><span class=s1>&#39;generated.plot.html&#39;</span><span class=p>]}</span>
</span></span></code></pre></div><h3 id=summary>Summary <a href=#summary class=anchor aria-hidden=true>#</a></h3><p>In this blog, we shared the experience of migrating a scRNA sequence analysis task written in R into Texera, and overcame the fact that Texera does not have built-in R UDF support yet. Specifically, we discussed how to use the <code>rpy2</code> library to execute R code in the Python environment, how to transfer R objects into the Python environment and how to transfer R objects between two operators on the workflow. We also discussed the design of decomposing the R script into a fully fledged workflow, with visualization enabled with HTML operators.</p><h4 id=acknowledgements>Acknowledgements <a href=#acknowledgements class=anchor aria-hidden=true>#</a></h4><p>Thanks to Prof. Shuibing Chen, Prof. Chen Li, and Dr. Tuo Zhang for their help with the task and the blog.</p></article></div></div></div></div><footer class="footer text-muted"><div class=container-fluid><div class=row><div class="col-lg-8 order-last order-lg-first"><ul class=list-inline><li class=list-inline-item></li></ul></div><div class="col-lg-8 order-first order-lg-last text-lg-end"><ul class=list-inline></ul></div></div></div></footer><script src=/js/bootstrap.min.0a0106334435d4110713bfaff4acaf02533b0c56d44412a74e9518db07672ddb7e099a9c7517bc9cb529e2ddd4ae2e438b40b720288900d5e74fb6d8928d9097.js integrity="sha512-CgEGM0Q11BEHE7+v9KyvAlM7DFbURBKnTpUY2wdnLdt+CZqcdRe8nLUp4t3Uri5Di0C3ICiJANXnT7bYko2Qlw==" crossorigin=anonymous defer></script>
<script src=/js/highlight.min.576fdb91cfd303e8fc247b90cc678e7d3304777de963077ad78eb70811a476919edaebf1d2843057250ec536b7ed1bd40f4e48f06e7cfc9c569dcf7581afd08b.js integrity="sha512-V2/bkc/TA+j8JHuQzGeOfTMEd33pYwd61463CBGkdpGe2uvx0oQwVyUOxTa37RvUD05I8G58/JxWnc91ga/Qiw==" crossorigin=anonymous defer></script>
<script src=/main.min.0382ba481d9dbcd1dd4c82111208ebf5d7eda2fce9378b0b14257d7f117c79f8ae01591f6ad0a99917b4e1624d8ace90de964e2431ab237000161ebd3c3dad63.js integrity="sha512-A4K6SB2dvNHdTIIREgjr9dftovzpN4sLFCV9fxF8efiuAVkfatCpmRe04WJNis6Q3pZOJDGrI3AAFh69PD2tYw==" crossorigin=anonymous defer></script></body></html>